drivers/hmis/app/models/hmis/ce/match/engine.rb

```diff
@@ -2,137 +2,157 @@
+      # optionally track the in-memory evaluations, which is the expensive work
+      progress_bar = nil
+      if progress
+        progress_bar = new_progress_bar
+        progress_bar.max += clients.count
+      end
 
-      now = DateTime.current
-      filtered_clients.in_batches do |batch|
-        # First iterate through the batch to import any Client Proxies that aren't present in the db already
-        proxies = []
-        batch.each do |client|
-          # In full mode, track clients as we process them
-          processed_client_ids.push(client.id) unless incremental
-          proxies << Hmis::Ce::ClientProxy.new(client: client)
+      started_at = Time.current
+      # SQL Prefiltering
+      prefilter_result = @prefilter.call(clients)
+
+      # remove any current clients that the sql filter excluded
+      if prefilter_result.lost_eligibility_clients.exists?
+        Hmis::Ce::Match::Candidate.transaction do
+          # capture snapshots at time of removal for the event log.
+          # The intent to log the attributes that caused the client to lose eligibility
+          snapshots = generate_snapshots(prefilter_result.lost_eligibility_clients, progress_bar)
+          @repo.remove_warehouse_client_candidates(prefilter_result.lost_eligibility_clients)
+          @event_writer.call(snapshots, timestamp: started_at)
         end
-        proxies_by_client = import_proxies!(proxies)
+      end
 
-        # Iterate through a second time to import candidate matches
-        candidates = []
+      prefilter_result.eligible_clients.in_batches do |batch|
+        now = Time.current
+        # import missing Client Proxies
+        @repo.import_proxies(batch, timestamp: now)
+        # efficient lookup for client_id => proxy_id
+        warehouse_proxy_map = @repo.proxy_by_warehouse_client(batch)
+        # track which clients in this batch are currently matched to the pool
+        current_warehouse_clients_ids = @repo.current_warehouse_client_ids(batch).to_set
+        # accumulate snapshots for changes to candidates
+        matching_client_snapshots = []
+        removed_client_snapshots = []
+
+        # Perform In-Memory Evaluation on each client
+        matching_candidates = []
         batch.each do |client|
-          bar&.increment!
-          # note, we could also set an expiration date on the candidate to allow us to skip records we have evaluated recently
-          next unless eligibility_evaluator.call(client)
+          progress_bar&.increment!
+          evaluation = @evaluator.call(client)
 
-          score = priority_evaluator.call(client)
-          # Client with nil score cannot be prioritized, so do not include them in the pool.
+          # Client without a score cannot be prioritized, so do not include them in the pool.
           # If needing to include clients that don't have a score, expression should be set up like `IF(my_score = NULL, 0, my_score)`
```

Comment:
This comment feels like it belongs in ClientPoolEvaluator now (or can be removed)